[DAG] Serve Deployment Graph documentation and tutorial. (#23512)
ray-project/ray: doc/source/serve/deployment-graph.md

---
jupytext:
  formats: ipynb,md:myst
  text_representation:
    extension: .md
    format_name: myst
    format_version: 0.13
    jupytext_version: 1.13.6
kernelspec:
  display_name: Python 3
  language: python
  name: python3
---

(serve-deployment-graph)=

# Deployment Graph

```{note} 
Note: This feature is still experimental in Alpha release, some APIs are subject to change.
```

> ## General Motivation

Reviewer: ```suggestion
## Motivation
```

> Production machine learning serving pipelines are getting longer and wider. They often consist of multiple, or even tens of models collectively making a final prediction, such as image / video content classification and tagging, fraud detection pipeline with multiple policies and models, multi-stage ranking and recommendation, etc.

Reviewer 2: ```suggestion
Production machine learning serving pipelines are getting longer and wider. They often consist of many models collectively making a final prediction. This is common in use cases like image / video content classification and tagging, fraud detection pipeline with multiple policies, multi-stage ranking and recommendation, etc.
```

> Meanwhile, the size of a model is also growing beyond the memory limit of a single machine due to the exponentially growing number of parameters, such as GPT-3, sparse feature embeddings in recsys models such that the ability to do disaggregated and distributed inference is desirable and future proof.

Reviewer 2: ```suggestion
Meanwhile, the size of a model is also growing beyond the memory limit of a single machine due to the exponentially growing number of parameters. GPT-3 and sparse feature embeddings in large recommendation models are two prime examples. The ability to do disaggregated and distributed inference is becoming more and more desirable.
```

> We want to leverage the programmable and general purpose distributed computing ability of Ray, double down on its unique strengths (scheduling, communication and shared memory) to facilitate authoring, orchestrating, scaling and deployment of complex serving pipelines under one set of DAG API, so a user can program & test multiple models or multiple shards of a single large model dynamically, deploy to production at scale, and upgrade individually.

Reviewer: ```suggestion
We want to leverage the programmable and general purpose distributed computing ability of Ray and double down on its unique strengths (scheduling, communication and shared memory) to facilitate authoring, orchestrating, scaling and deployment of complex serving graphs so a user can program & test multiple models or multiple shards of a single large model dynamically and deploy to production at scale.
```

Reviewer: we don't really support upgrading individually yet, that can come later :)

Author: we can rephrase it as individually scalable (# of replicas) and reconfigurable (model weight path), both pretty valuable and probably covers 95% of the actual upgrade needs that other frameworks don't have.

## Key requirements
- Provide the ability to author a DAG of Serve nodes to form a complex inference graph.

> - Pipeline authoring experience should be fully Python-programmable with support for dynamic selection, control flows, user business logic, etc.

Reviewer: "Pipeline" should be "Graph" or "DAG" to match others

> - DAG can be instantiated and locally executed using tasks and actors API

Reviewer: suggest to remove this bullet, it doesn't make much sense in the context of these docs

Author: how about we rename it to "Key Features" ? I'm putting this as a minimal list of our P0 design requirements that led to what it is now, without requiring user to dig deep into the examples to understand. User should be able to make a quick call after reading this and think "ok this is what i'm looking for"

- DAG can be deployed as a group where individual nodes can be reconfigured and scaled indepenently.

__[Full Ray Enhancement Proposal, REP-001: Serve Pipeline](https://github.com/ray-project/enhancements/blob/main/reps/2022-03-08-serve_pipeline.md)__

+++

## Concepts

### Deployment

> Upgradeable group of actors managed by the Serve controller. Currently the primary API in Ray Serve. At authoring time it’s the class or function under `@serve.deployment` decorator. 

Reviewer: ```suggestion
Scalable, upgradeable group of actors managed by Ray Serve. 
```

Reviewer: link to the deployment docs section, no need to re-write that here

### Node

> Smallest unit in a graph, typically an annotated class but can also be a function, backed by a group of actors that are scalable and reconfigurable. 

Reviewer: this is confusing, it doesn't explain that it's 1-1 with a deployment. How about combine the two and call it "Deployment Node" ?

### Deployment Graph

> Collection of nodes that forms a DAG that represents an inference graph for complicated tasks. Ex: ensemble, chaining, dynamic selection. 

Reviewer 2: ```suggestion
Collection of nodes that forms a DAG that represents an inference graph for complicated tasks. For example: ensemble, chaining, dynamic selection. 
```

### Bind

> A graph building API applicable to decorated class or function.  `decorated_class_or_func.bind(*args, **kwargs)` generates an IR node that can be used to build graph, and bound arguments will be applied at execution time, including dynamic user input.

Reviewer: nit: reorder these so they flows conceptually:

- DeploymentNode: single node in a graph
- Bind: used to bind multiple nodes together
- Deployment Graph: multiple deployments bound together that can be deployed and call as a unit

+++

## Key APIs Explained

The class and function definition as well as decorator didn't diverage from existing serve API. So in the following section we only need to dive into a few new key APIs used: `bind()`, `InputNode()` and `serve.run()`

+++

### **`bind(*args, **kwargs)`**

> Once called on supported ray decorated function or class (@ray.remote, @serve.deployment), generates an IR of type DAGNode that acts as the building block of graph building.

Reviewer 2: ```suggestion
Once called on supported ray decorated function or class (``@serve.deployment`` is fully supported right now, ``@ray.remote`` coming soon), generates an IR of type DAGNode that acts as the building block of graph building.
```

We need to write end to end examples of ray.remote before promising fully working :D

+++

#### On function

```bind()``` on function produces a DAGNode that can be exeucted with user input.

```{code-cell} ipython3
import ray
from ray import serve

> from ray.serve.pipeline.generate import DeploymentNameGenerator

Reviewer: this should not be user exposed

Reviewer 2: +1 I understand we need this for notebook for your local dev, but it needs to be cleaned up users shouldn't need to see this!

Author: explained below

if ray.is_initialized():
    serve.shutdown()
    DeploymentNameGenerator.reset()
    ray.shutdown()

ray.init(num_cpus=4)

> serve.start()

Reviewer: do we need all of this boilerplate?

Reviewer: can't users just do `serve run` repeatedly?

Author: explained below

### Setting up clean ray cluster with serve ###

@serve.deployment
def preprocessor_with_arg(val):
    print(val)

# Produces a DeploymentFunctionNode with no bound args.
func_node = preprocessor_with_arg.bind()

dag_handle = serve.run(func_node)
# Once executed it will execute with arg value of 2.
print(ray.get(dag_handle.remote(2)))

"""
(my_func pid=17060) 2
"""
```

#### On class constructor 

> **`Class.bind(*args, **kwargs)`** constructs and returns a DAGNode that acts as the instantiated instance of Class, where `*args` and `**kwargs` are used as init args.

Reviewer: let's also explain that a bound node can be passed to another, and this will be resolved at runtime to a callable serve handle

Reviewer: Would suggest adding a short code snippet here, building up incrementally.

#### On class method

Once a class is bound with its init args, its class methods can be directly accessed, called or bound with other args.

```{code-cell} ipython3
import ray
from ray import serve
from ray.serve.pipeline.generate import DeploymentNameGenerator

if ray.is_initialized():
    serve.shutdown()
    DeploymentNameGenerator.reset()
    ray.shutdown()

ray.init(num_cpus=4)

> serve.start()

Reviewer: don't think we need any of this, just `serve run` from CLI?

> ### Setting up clean ray cluster with serve ###

Reviewer: what does clean mean? commentary should generally be in docs rather than comments also

Author: i found the restarting and resetting serve necessary in jupyter notebook setting for users, even when i repeated run the notebook myself. Two things ..

1) If we run multiple graph in one notebook, second one fails because it's trying to take `/` route prefix by default
2) DeploymentNameGenerator as a class level import didn't got cleaned up after serve.shutdown (we should fix this) that will lead to problems if user play around this documentation more than one time since it will spawn new replicas with new names but user laptop will run out of cores to host everything.

> @serve.deployment

Reviewer: ditto with the above

class Model:
    def __init__(self, val):
        self.val = val
    def get(self):
        return self.val

# Produces a deployment class node instance initialized with bound args value. 
class_node = Model.bind(5)
dag_handle = serve.run(class_node)
# Access get() class method on bound Model
print(ray.get(dag_handle.get.remote()))
"""
get() returns

5
"""
```

> ### DAGNode as args in bind()

Reviewer 2: ```suggestion
### DAGNode as arguments in other bind()
```

DAGNode can also be passed into other DAGNode in dag binding. In the full example below, ```Combiner``` calls into two instantiations of ```Model``` class, which can be bound and passed into ```Combiner```'s constructor as if we're passing in two regular python class instances.

```python
m1 = Model.bind(1)
m2 = Model.bind(2)
combiner = Combiner.bind(m1, m2)
```

Similarly, we can also pass and bind upstream DAGNode results that will be resolved upon runtime to downstream DAGNodes, in our example, a `DeploymentMethodNode` that access class method of ```Combiner``` class takes two preprocessing DAGNodes' output as well as part of user input.

```python

> preprocessed_1 = preprocessor.bind(dag_input[0])

Reviewer 2: you haven't introduced dag input yet, maybe switch the ordering?

Author: the ordering is a bit unnatural to me now ... i started off the example first then positioned all other cells down the line to explain what happened. I think now it makes much more sense to just move it back to top but build it up in the incremental way.

preprocessed_2 = avg_preprocessor.bind(dag_input[1])
...
dag = combiner.run.bind(preprocessed_1, preprocessed_2, dag_input[2])
```

+++

### **```InputNode()```** : User input of the graph

> ```InputNode``` is a special singleton in DAG building that's only relevant to it's runtime call behavior. Even though all decorated classes or functions can be reused in arbitrary way to facilitate DAG building where the root DAGNode forms the graph with its children, in each deployment graph there should be one and only one InputNode used.

Reviewer 2: need example of it and mention it has to be used as context manager

```

+++

### **```serve.run()```** : running the deployment graph

The deployment graph can be deployed with ```serve.run()```. ```serve.run()```
takes in a target DeploymentNode, and it deploys the node's deployments, as
well as all its child nodes' deployments. To deploy your graph, pass in the
driver DeploymentNode into ```serve.run()```:

```python
with InputNode() as dag_input:
    # ... DAG building
    serve_dag = ...
dag_handle = serve.run(serve_dag)
```

```serve.run``` returns the passed-in node's deployment's handle. You can use
this handle to issue requests to the deployment:

```python
# Warm up
ray.get(dag_handle.predict.remote(["0", [0, 0], "sum"]))
```

During development, you can also use the Serve CLI to run your deployment
graph. The CLI was included with Serve when you did ``pip install "ray[serve]"``.
The command ```serve run [node import path]``` will deploy the node and its
childrens' deployments. For example, we can remove the ```serve.run()``` calls
inside the Python script and save our example pipeline to a file called
example.py. Then we can run the driver DeploymentNode using its import path,
```example.serve_dag```:

```bash

> $ ray start --head

Reviewer: don't need `ray start` if using `serve run`

    time.sleep(0.15) # Manual delay for blocking computation
    return sum(input_data) / len(input_data)

@serve.deployment
class Model:
    def __init__(self, weight: int):
        self.weight = weight

    async def forward(self, input: int):
        time.sleep(0.3) # Manual delay for blocking computation 
        return f"({self.weight} * {input})"

@serve.deployment
class Combiner:
    def __init__(self, m1: Model, m2: Model):
        self.m1 = m1
        self.m2 = m2

    async def run(self, req_part_1, req_part_2, operation):
        # Merge model input from two preprocessors  
        req = f"({req_part_1} + {req_part_2})"
        
        # Submit to both m1 and m2 with same req data in parallel
        r1_ref = self.m1.forward.remote(req)
        r2_ref = self.m2.forward.remote(req)
        
        # Async gathering of model forward results for same request data
        rst = await asyncio.gather(*[r1_ref, r2_ref])
        
        # Control flow that determines runtime behavior based on user input
        if operation == "sum":
            return f"sum({rst})"
        else:
            return f"max({rst})"
        
@serve.deployment(num_replicas=2)

> class DAGDriver:

Reviewer 2: let's just use the built-in driver now?

Author: i tried but i didn't work due to the input_schema thing :/  it also some some rough edges of user input because ray dag's InputNode accepts .remote(1, 2, x=3) but it didn't match default args of DAGDriver.

Author: i had it earlier while putting up the example: https://github.com/ray-project/ray/issues/23484

+++

## Concepts

### Deployment
Scalable, upgradeable group of actors managed by Ray Serve. __[See docs for detail](https://docs.ray.io/en/master/serve/core-apis.html#core-api-deployments)__

### DeploymentNode
Smallest unit in a graph, typically a serve annotated class or function, backed by a Deployment.

### Bind
A graph building API applicable to decorated class or function.  `decorated_class_or_func.bind(*args, **kwargs)` generates an IR node that can be used to build graph, and bound arguments will be applied at execution time, including dynamic user input.

### Deployment Graph
Collection of deployment nodes bound together that forms a DAG that represents an inference graph for complicated tasks, can be deployed and call as a unit. Ex: ensemble, chaining, dynamic selection. 

+++

## Key APIs Explained

The class and function definition as well as decorator didn't diverage from existing serve API. So in the following section we only need to dive into a few new key APIs used: `bind()`, `InputNode()` and `serve.run()`

+++

### **`bind(*args, **kwargs)`**

> Once called on supported ray decorated function or class (@serve.deployment fully supported, @ray.remote soon), generates an IR of type DAGNode that acts as the building block of graph building.

Reviewer 3: ```suggestion
Once called on supported Ray-decorated function or class (`@serve.deployment` is fully supported, `@ray.remote` will be soon), generates an IR of type DAGNode that acts as the building block of graph building.
```

```InputNode``` value is fulfilled and replaced by user input at runtime, therefore it takes no argument when being constructed.

It's possible to access partial user input by index or key, if some DAGNode in the graph doesn't need the complete user input to run. In the full example below, `combiner.run` only needs the element at index 2 to determine it's runtime behavior.

```python
dag = combiner.run.bind(preprocessed_1, preprocessed_2, dag_input[2])
```

+++

### **```serve.run()```** : running the deployment graph

The deployment graph can be deployed with ```serve.run()```. ```serve.run()```
takes in a target DeploymentNode, and it deploys the node's deployments, as
well as all its child nodes' deployments. To deploy your graph, pass in the
driver DeploymentNode into ```serve.run()```:

```python
with InputNode() as dag_input:
    # ... DAG building
    serve_dag = ...
dag_handle = serve.run(serve_dag)
```

> ```serve.run``` returns the passed-in node's deployment's handle. You can use

Reviewer 3: ```suggestion
```serve.run()``` returns the passed-in node's deployment's handle. You can use
```

```{tip}
At the end of this document we have the full and end to end executable implementation.
```

- All nodes in the deployment graph naturally forms a DAG structure.
- A node could use or call into other nodes in the deployment graph.
- Deployment graph has mix of class and function as nodes.
- Same input or output can be used in multiple nodes in the DAG.
- A node might access partial user input.
- Control flow is used where dynamic dispatch happens with respect to input value.
- Same class can be constructed, or function bound with different args that generates multiple distinct nodes in DAG.
- A node can be called either sync or async.

+++

![deployment graph](https://github.com/ray-project/images/blob/master/docs/serve/deployment_graph.png?raw=true)

+++

### Step 1: User InputNode and preprocessor 

Let's start with the first layer of DAG: Building user input to two preprocessor functions, where each function receives parts of the input data. For simplicity, we use the same existing `@serve.deployment` decorator on an async function body.

+++

```python
from ray import serve
from ray.experimental.dag.input_node import InputNode

@serve.deployment

> async def preprocessor(input_data: str):

Author: i want to show an async example with blocking computations that user might not expect, then either do `asyncio.gather`  or showing user can directly bind an async function. I think alternatively i can do `await asyncio.sleep()` to yield control but im more in favor of the current one since user might not put heavy computation with right `await` syntax in actual implementation.

> async def preprocessor(input_data: str):

Reviewer 2: +1 for await asyncio. you can show the async def somewhere else (in particular the driver node)

> async def preprocessor(input_data: str):

Author: updated

    """Simple feature processing that converts str to int"""
    time.sleep(0.1) # Manual delay for blocking computation
    return int(input_data)

@serve.deployment

> async def avg_preprocessor(input_data):

Author: ditto

    time.sleep(0.15) # Manual delay for blocking computation
    return sum(input_data) / len(input_data)

# DAG building
with InputNode() as dag_input:
    preprocessed_1 = preprocessor.bind(dag_input[0])  # Partial access of user input by index
    preprocessed_2 = avg_preprocessor.bind(dag_input[1]) # Partial access of user input by index
```

+++

There are two new APIs used in the DAG building stage: `InputNode()` and `bind()`.

### **```InputNode()```** : User input of the graph

```InputNode``` is a special singleton in DAG building that's only relevant to it's runtime call behavior. Even though all decorated classes or functions can be reused in arbitrary way to facilitate DAG building where the root DAGNode forms the graph with its children, in each deployment graph there should be one and only one InputNode used.

```InputNode``` value is fulfilled and replaced by user input at runtime, therefore it takes no argument when being constructed.

```InputNode``` should only be created within a context manager as syntax sugar to indicate there's one and only one `InputNode` for each graph, and will throw exception if directly created, or multiple is found.

> It's possible to access partial user input by index or key, if some DAGNode in the graph doesn't need the complete user input to run. 

Author: i've updated it to 

```
It's possible to access partial user input by index or key, if a node in the graph doesn't need the complete user input to run. Example: User submits `[Tensor_0, Tensor_1]` via http but a model node might only need to receive `Tensor_1` at index 1 rather than complete list to reduce intermediate data transfer.
```

    ...

    async def run(self, req_part_1, req_part_2, operation):
        # Merge model input from two preprocessors  
        req = f"({req_part_1} + {req_part_2})"
        
        # Submit to both m1 and m2 with same req data in parallel
        r1_ref = self.m1.forward.remote(req)
        r2_ref = self.m2.forward.remote(req)
        
        # Async gathering of model forward results for same request data
        rst = await asyncio.gather(*[r1_ref, r2_ref])
        
        # Control flow that determines runtime behavior based on user input
        if operation == "sum":
            return f"sum({rst})"
        else:
            return f"max({rst})"
```

+++

### Step 4: Driver deployment to handle http ingress

Now we've built the entire serve DAG with the topology, args binding and user input. It's time to add the last piece for serve -- a Driver deployment to expose and configure http. We can configure it to start with two replicas in case the ingress of deployment becomes bottleneck of the DAG.

```{note}
We expect each DAG has a driver class implementation as root, similar to the example below. This is where HTTP ingress are configured and implemented. 

> ```

Reviewer 2: can you please mention the default http driver? 
and mention this is for advanced only. 
and open an issue for how it currently cannot be support.

Author: sg i think we have the issue there already that you're working on, let me add the part for a default DAGDriver.

Author: my only concern is unless we have your fix cherrypicked this notebook won't run with default DAGDriver :/

Reviewer 2: we can keep this, just need to rework once default driver is fixed.

Author: addressed in latest commit -- please let me know if you think the content looks good !

Reviewer 2: looks good for now!

---
jupytext:
  formats: ipynb,md:myst
  text_representation:
    extension: .md
    format_name: myst
    format_version: 0.13
    jupytext_version: 1.13.6
kernelspec:
  display_name: Python 3
  language: python
  name: python3
---

(serve-deployment-graph)=

# Deployment Graph

```{note} 
Note: This feature is in Alpha, so APIs are subject to change.
```

## Motivation

> Production machine learning serving pipelines are getting longer and wider. They often consist of many models collectively making a final prediction. This is common in use cases like image / video content classification and tagging, fraud detection pipeline with multiple policies, multi-stage ranking and recommendation, etc.

Reviewer: ```suggestion
Machine learning serving systems often consist of many models to make a single prediction. This is common in use cases like image / video content classification and tagging, fraud detection pipeline with multiple policies, multi-stage ranking and recommendation, etc.
```

Author: hmm i see what you mean here but the user call I had with Zilliz, their primary POC literally says he cannot agree more with the opening statement of "Production machine learning serving pipelines are getting longer and wider" that matches exactly the problems they're going through. How about i make your suggested edits of the paragraph but keep the first line ?

Reviewer: that's fine

> Meanwhile, the size of a model is also growing beyond the memory limit of a single machine due to the exponentially growing number of parameters. GPT-3 and sparse feature embeddings in large recommendation models are two prime examples. The need of serving large models with disaggregated and distributed inference is rapidly growing.

Reviewer: ```suggestion
Meanwhile, the size of a model is also growing beyond the memory limit of a single machine due to the exponentially growing number of parameters. GPT-3 and sparse feature embeddings in large recommendation models are two prime examples.
```

> We want to leverage the programmable and general purpose distributed computing ability of Ray and double down on its unique strengths (scheduling, communication and shared memory) to facilitate authoring, orchestrating, scaling and deployment of complex serving graphs so a user can program & test multiple models or multiple shards of a single large model dynamically and deploy to production at scale, and able to scale and reconfigure individually.

Reviewer: ```suggestion
Ray has unique strengths suited to distributed inference pipelines: flexible scheduling, efficient communication, and shared memory. Ray Serve leverages these strengths to build inference graphs, enabling users to develop complex ML applications locally and then deploy them to production with dynamic scaling and lightweight updates (e.g., for model weights).
```

- Graph authoring experience should be fully Python-programmable with support for dynamic selection, control flows, user business logic, etc.
- Graph can be instantiated and locally executed using tasks and actors API
- Graph can be deployed as a group where individual nodes can be reconfigured and scaled indepenently.
- User should not need to write any YAML file to author, run, iterate, and deploy a graph. 
- Have a path for DevOps / MLOps people to take actions on operationalizing a deployment graph with minimal knowledge required of the code.

__[Full Ray Enhancement Proposal, REP-001: Serve Pipeline](https://github.com/ray-project/enhancements/blob/main/reps/2022-03-08-serve_pipeline.md)__

+++

## Concepts

- **Deployment**: Scalable, upgradeable group of actors managed by Ray Serve. __[See docs for detail](https://docs.ray.io/en/master/serve/core-apis.html#core-api-deployments)__

- **DeploymentNode**: Smallest unit in a graph, typically a serve annotated class or function, backed by a Deployment.

- **InputNode**: Singleton node in a graph that holds user input value, which is resolved at runtime of graph execution.

> - **Bind**: A graph building API applicable to decorated class or function.  `decorated_class_or_func.bind(*args, **kwargs)` generates a `DeploymentNode` that can be used to build graph, and bound arguments will be applied at execution time, including dynamic user input.

Author: sg

+++

## Full End to End Example Walkthrough

Let's put the concepts together and incrementally build a runnable DAG example highlighting the following features:

```{tip}
At the end of this document we have the full and end to end executable implementation.
```

- Building a graph:
    - A deployment node is created from `@serve.deployment` decorated function or class. 
    - You can construct different deployment nodes from same class or function.
    - Deployment nodes can be used as input args in .bind() of other nodes in the DAG. 
    - Multiple nodes in the deployment graph naturally forms a DAG structure.
- Accessing input:
    - Same input or output can be used in multiple nodes in the DAG.
    - Deployment nodes can access partial user input.
- Dynamic control flow:
    - A deployment node can call into other nodes in the deployment graph.
    - You can use the dynamic calling to perform control flow operation that's hard to expressive in traditional DAG.
- Running a DAG:

>     - A node can be called either sync or async.

Author: oh i mean the deployment function or class method can be sync or async, just to show that in deployment graph execution we didn't break or enforce anything and it works as expected even out of order.

>     - A node can be called either sync or async.

Author: i changed it to `    - Nodes in the graph, such as functions or class methods, can be either sync or async.`

>     - A node can be called either sync or async.

Reviewer: ah ok, updated version is more clear thanks

    time.sleep(0.1) # Manual delay for blocking computation
    return int(input_data)

@serve.deployment
async def avg_preprocessor(input_data):
    """Simple feature processing that returns average of input list as float."""
    time.sleep(0.15) # Manual delay for blocking computation
    return sum(input_data) / len(input_data)

# DAG building
with InputNode() as dag_input:
    # Partial access of user input by index
    preprocessed_1 = preprocessor.bind(dag_input[0])  
    preprocessed_2 = avg_preprocessor.bind(dag_input[1])
```

+++

There are two new APIs used in the DAG building stage: `InputNode()` and `bind()`.

### **```InputNode()```** : User input of the graph

```InputNode``` is a special singleton in DAG building that's only relevant to it's runtime call behavior. Even though all decorated classes or functions can be reused in arbitrary way to facilitate DAG building where the root DAGNode forms the graph with its children, in each deployment graph there should be one and only one InputNode used.

```InputNode``` value is fulfilled and replaced by user input at runtime, therefore it takes no argument when being constructed.

> ```InputNode``` should only be created within a context manager to indicate there's one and only one `InputNode` for each graph, and will throw exception if directly created, or multiple is found.

Author: i can do the shortening, but i don't think `InputNode` is the root node, i think it's the opposite leaf node that always gets resolved first when we try to execute the DAG. From user perspective, the root node is the entrypoint node of user choice, such as `combine.run()`; For execution, `InputNode` gets replaced by user value in first layer. So for user facing docs I think we probably want to let user think InputNode is just your inputs in arg rather than root of the  dag.

It's possible to access partial user input by index or key, if a node in the graph doesn't need the complete user input to run. Example: User submits `[Tensor_0, Tensor_1]` via http but a model node might only need to receive `Tensor_1` at index 1 rather than complete list to reduce intermediate data transfer.

### **`bind(*args, **kwargs)`** : The graph building API

Once called on supported Ray-decorated function or class (`@serve.deployment` is fully supported, `@ray.remote` will be soon), generates a `DeploymentNode` of type `DAGNode` that acts as the building block of graph building.

> 

Reviewer: this seems redundant with the concepts section above, can we combine them into one? just incrementally introduce the concepts?

Author: combined by removing bind from concepts section

In the example above, we can see we're using a context manager to build and bind user input:
```python
with InputNode() as dag_input:
```

Which can be used and accessed by index or key in downstream calls of .bind(), such as:
```python
preprocessed_1 = preprocessor.bind(dag_input[0])
```

> This means we're creating a DAGNode IR called `preprocessed_1` in graph building by calling `.bind()` on a serve decorated function, where it executes the decorated deployment function `preprocessor` that takes the user input at index 0 at runtime.

Reviewer: `DAGNode` and `IR` are not concepts that have been introduced

> This means we're creating a DAGNode IR called `preprocessed_1` in graph building by calling `.bind()` on a serve decorated function, where it executes the decorated deployment function `preprocessor` that takes the user input at index 0 at runtime.

Reviewer: ```suggestion
This creates a deployment node called `preprocessed_1` by calling `.bind()` on the decorated function. Once deployed, this will translate to a deployment that runs `preprocessor` on the input at runtime.
```

#### bind() on function

```bind()``` on function produces a DAGNode that can be executed with user input.

```{tip}

> Each deployment node used in graph is individually scalable and configurable by default. This means in real production workload where we can expect difference in compute resource and latency, we can fine tune the nodes to optimal `num_replicas` and `num_cpus` to avoid a single node being the bottleneck of your deployment graph's latency or throughput.

Reviewer: we should have a performance tuning section in the future!

Author: this is the beginning of where things get interesting ;)

        r1_ref = self.m1.forward.remote(req)
        r2_ref = self.m2.forward.remote(req)
        
        # Async gathering of model forward results for same request data
        rst = await asyncio.gather(*[r1_ref, r2_ref])

# DAG building
with InputNode() as dag_input:
    preprocessed_1 = preprocessor.bind(dag_input[0])  # Partial access of user input by index
    preprocessed_2 = avg_preprocessor.bind(dag_input[1]) # Partial access of user input by index
    m1 = Model.bind(1)
    m2 = Model.bind(2)
    combiner = Combiner.bind(m1, m2)
    dag = combiner.run.bind(
        preprocessed_1, preprocessed_2, dag_input[2]  # Partial access of user input by index
    ) 
```

+++

We are adding a few more pieces to our dag builder: `bind()` on class and class method, as well as passing the output of `Model.bind()` as init args into another class `Combiner.bind()`

#### bind() on class constructor 

**`Class.bind(*args, **kwargs)`** constructs and returns a DAGNode that acts as the instantiated instance of Class, where `*args` and `**kwargs` are used as init args. In our implementation, we have

```python
m1 = Model.bind(1)
m2 = Model.bind(2)
```

> This means we're creating two more DAGNodes of an instance of `Model` that is constructed with init arg of `1` and `2`, and refereced with variable name `m1`, `m2` respectively.

Reviewer: still think this should be combined with the part about functions, they're basically the same thing except classes take args

Author: moved, seem like it didn't hurt reading experience when we introduce passing deployment node as args

#### bind() on class method

Once a class is bound with its init args, its class methods can be directly accessed, called or bound with other args. It has the same semantics as `bind()` on a function, except it acts on an instantiated `DAGNode` class instead.

### DAGNode as arguments in other node's bind()

DAGNode can also be passed into other DAGNode in dag binding. In the full example below, ```Combiner``` calls into two instantiations of ```Model``` class, which can be bound and passed into ```Combiner```'s constructor as if we're passing in two regular python class instances.

```python
m1 = Model.bind(1)
m2 = Model.bind(2)
combiner = Combiner.bind(m1, m2)
```

> Similarly, we can also pass and bind upstream DAGNode results that will be resolved upon runtime to downstream DAGNodes, in our example, a function `run()` that access class method of ```Combiner``` class takes two preprocessing DAGNodes' output as well as part of user input that will be resolved when upstream `DAGNode` are executed.

Reviewer: `DAGNode` was never introduced but you're using it in a few different places, please call them deployment nodes to be consistent.

```{note}
We expect each DAG has a driver class implementation as root, similar to the example below. This is where HTTP ingress are configured and implemented. We provide a default `DAGDriver` to handle simple HTTP parsing, but in this example we put up a custom implementation. 
```

+++

```python
@serve.deployment(num_replicas=2)
class DAGDriver:
    def __init__(self, dag_handle):
        self.dag_handle = dag_handle

    async def predict(self, inp):
        """Perform inference directly without HTTP."""
        return await self.dag_handle.remote(inp)

    async def __call__(self, request: starlette.requests.Request):
        """HTTP endpoint of the DAG."""
        input_data = await request.json()
        return await self.predict(input_data)

# DAG building
with InputNode() as dag_input:
    ...
    dag = combiner.run.bind(
        preprocessed_1, preprocessed_2, dag_input[2]  # Partial access of user input by index
    ) 
    
    # Each serve dag has a driver deployment as ingress that can be user provided.
    serve_dag = DAGDriver.options(route_prefix="/my-dag").bind(dag)
```

+++

### Step 5: Test the full DAG in both python and http

> We can then try to deploy and run the steps we had above on a simple function DAG, using our new API `serve.run()`.

Reviewer: ```suggestion
You can test and deploy the graph using `serve.run`:
```

- Accessing input:
    - Same input or output can be used in multiple nodes in the DAG.
    - Deployment nodes can access partial user input.
- Dynamic control flow:
    - A deployment node can call into other nodes in the deployment graph.
    - You can use the dynamic calling to perform control flow operation that's hard to expressive in traditional DAG.
- Running a DAG:
    - Nodes in the graph, such as functions or class methods, can be either sync or async.

+++

![deployment graph](https://github.com/ray-project/images/blob/master/docs/serve/deployment_graph.png?raw=true)

+++

### Step 1: User InputNode and preprocessor 

Let's start with the first layer of DAG: Building user input to two preprocessor functions, where each function receives parts of the input data. For simplicity, we use the same existing `@serve.deployment` decorator on an async function body.

+++

```python
from ray import serve
# We will later move Ray DAG related components 
# out of experimental in later stable release
from ray.experimental.dag.input_node import InputNode

@serve.deployment
async def preprocessor(input_data: str):
    """Simple feature processing that converts str to int"""

>     time.sleep(0.1) # Manual delay for blocking computation

Reviewer: still looks like this is `time.sleep` ?

Author: ah i forgot to update this one. pushing new commit now
